tiledb/cloud/dag/dag.py

```diff
@@ -777,28 +963,36 @@ def wait(self, timeout=None):
                 "timeout must be numeric value representing seconds to wait"
             )
 
-        start_time = time.time()
-        end_time = None
-        if timeout is not None:
-            end_time = start_time + timeout
-        while not self.done():
-            if end_time is not None and time.time() >= end_time:
-                raise futures.TimeoutError(
-                    "timeout of {} reached and dag is not complete".format(timeout)
-                )
-            time.sleep(0.01)
+        with self._lifecycle_condition:
+            futures.wait_for(self._lifecycle_condition, self._done, timeout)
 
         # in case of failure reraise the first failed node exception
         if self.status == st.Status.FAILED:
-            raise next(iter(self.failed_nodes.values())).error
+            exc = next(iter(self.failed_nodes.values())).error
+            assert exc
+            raise exc
 
     def cancel(self):
-        if self.status not in (st.Status.RUNNING, st.Status.NOT_STARTED):
-            return
-        for node in list(self.running_nodes.values()):
-            node.cancel()
-        for node in list(self.not_started_nodes.values()):
-            node.cancel()
+        with self._lifecycle_condition:
```

Comment:
Not sure if this is in the scope of this PR but some functions annotations like this one missing the return type. If no please ignore.

Reply:
I've been adding annotations to functions in an ad-hoc manner over the course of working on this module. It admittedly would probably be good to add them to everything but it's not a priority.